tests/e2e-evm/test/callcode.test.ts

```diff
@@ -0,0 +1,247 @@
+import hre from "hardhat";
+import type { ArtifactsMap } from "hardhat/types/artifacts";
+import type { PublicClient, WalletClient, GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";
+import { expect } from "chai";
+import { Address, CallParameters, Chain, decodeFunctionResult, encodeFunctionData, getAddress, pad, toHex } from "viem";
+import { whaleAddress } from "./addresses";
+
+const defaultGas = 25000n;
+const contractCallerGas = defaultGas + 10000n;
+
+describe("CallCode", () => {
+  //
+  // Client + Wallet Setup
+  //
+  let publicClient: PublicClient;
+  let walletClient: WalletClient;
+  let lowLevelCaller: GetContractReturnType<ArtifactsMap["Caller"]["abi"]>;
+  before("setup clients", async function () {
+    publicClient = await hre.viem.getPublicClient();
+    walletClient = await hre.viem.getWalletClient(whaleAddress);
+    lowLevelCaller = await hre.viem.deployContract("Caller");
+  });
+
+  interface CallCodeContext {
+    lowLevelCaller: GetContractReturnType<ArtifactsMap["Caller"]["abi"]>;
+    implementationContract: GetContractReturnType<ArtifactsMap["CallCodeTestContract"]["abi"]>;
+  }
+
+  interface callCodeTestCaseBase {
+    name: string;
+    txParams: (ctx: CallCodeContext) => CallParameters<Chain>;
+  }
+
+  // callCodeTestCaseSendAndValue is for getMsgInfo() call tests to validate
+  // expected msg.sender and msg.value
+  type callCodeTestCaseSendAndValue = callCodeTestCaseBase & {
+    wantSender: (ctx: CallCodeContext, signer: Address) => Address;
+    wantValue: bigint;
+    wantStorageContract?: never;
+    wantStorageValue?: never;
+  };
+
+  // callCodeTestCaseStorage is for setStorageValue() call tests to validate
+  // which contract the storage is set on and the expected storage value
+  type callCodeTestCaseStorage = callCodeTestCaseBase & {
+    wantSender?: never;
+    wantValue?: never;
+    wantStorageContract: (ctx: CallCodeContext) => Address;
+    wantStorageValue: bigint;
+  };
+
+  type callCodeTestCase = callCodeTestCaseSendAndValue | callCodeTestCaseStorage;
```

Comment:
This is quite a nice method for adding multiple kinds of tests in the same suite, thanks for teaching me this ✍️

Reply:
The `want*` prefix is surprising nice to read